### PR TITLE
[DMs] Update DMs store, sagas and selectors to be optimistic

### PR DESCRIPTION
**Pull request by rickyrombo**
### Description

- Makes sending messages, reactions, and marking chats as read optimistic, giving the illusion of speed
- Undoes the optimistic changes if request fails
- Adds a saga that fetches chats if messages are being added from a chat we haven't fetched yet

### Dragons

*Is there anything the reviewer should be on the lookout for? Are there any dangerous changes?*

### How Has This Been Tested?

*Please describe the tests that you ran to verify your changes. Provide repro instructions & any configuration.*

### How will this change be monitored?

*For features that are critical or could fail silently please describe the monitoring/alerting being added.*

### Feature Flags ###

*Are all new features properly feature flagged? Describe added feature flags.*

### CODE REVIEW

**Comment by audius-infra** [comment]
Preview this change https://demo.audius.co/mjp-dms-optimism

File: packages/common/src/store/pages/chat/sagas.ts [hunk 1]
```diff
@@ -133,29 +133,51 @@ function* doMarkChatAsRead(action: ReturnType<typeof markChatAsRead>) {
...
-      }
+      yield* put(markChatAsReadSucceeded({ chatId }))
     }
   } catch (e) {
     console.error('markChatAsReadFailed', e)
+    yield* put(markChatAsReadFailed({ chatId }))
   }
 }
 
+let tempMessageIdCounter = 1
```
Comment (rickyrombo) [inline packages/common/src/store/pages/chat/sagas.ts:144]: Will probably iterate on this - might have SDK expose the cuid() generation and use that... but this works for now...

File: packages/common/src/store/pages/chat/selectors.ts [hunk 2]
```diff
@@ -13,23 +15,80 @@ export const getChat = (state: CommonState, chatId?: string) =>
...
 
+export const getOptimisticReactions = (state: CommonState) =>
+  state.pages.chat.optimisticReactions
+
+export const getChats = createSelector(
+  [getChatsRaw, getAllChatMessages, getOptimisticReads],
+  (chats, allMessages, optimisticReads) => {
+    return chats?.map((chat) => {
+      const chatMessages = allMessages[chat.chat_id]?.data
+      if (
```
Comment (piazzatron) [inline packages/common/src/store/pages/chat/selectors.ts:50]: Maybe a comment here explaining this logic

**Review by piazzatron — APPROVED** [review]

File: packages/common/src/store/pages/chat/slice.ts [hunk 3]
```diff
@@ -126,52 +141,99 @@ const slice = createSlice({
...
+      state,
+      action: PayloadAction<SetMessageReactionPayload>
+    ) => {
+      // Reset our optimism :(
+      const { messageId } = action.payload
+      delete state.optimisticReactions[messageId]
+    },
+    fetchChatSucceeded: (state, action: PayloadAction<{ chat: UserChat }>) => {
+      const { chat } = action.payload
+      if (!state.chatList.data.find((c) => c.chat_id === chat.chat_id)) {
```
[the remark this answers is not in the document]
Comment (rickyrombo) [reply packages/common/src/store/pages/chat/slice.ts:154]: yeah... so far i've been keeping the store pretty "dumb" but I think we could get some small perf wins with more cleverness and maps

**Pull request merged**